python/addPython.go

```diff
@@ -0,0 +1,142 @@
+		return false
+	}
+	fmt.Printf("%s found in path\n", toolName)
+	return true
+}
+
+// Creates a conda environment called projName
+// Conda must be installed for this to work
+// * tested
+func createCondaEnv(projName string) error {
+	err := exec.Command("conda", "create", "-n", projName).Run()
+	if err != nil {
+		fmt.Println(err.Error())
+		return err
+	}
+	fmt.Println("conda environment created: " + projName)
+	return nil
+}
+
+// This is a very hacky solution but conda has a problem where you can't
+// Easily use conda activate in subshells. Because go execute commands run
+// in a subshell, it's hard to circumvent this. There are workarounds for
+// bash scripts, but none were easily portable to a go function. So for now,
+// this creates and deletes a shell file that uses the workarounds.
+// ? Look for better solutions
+// * tested
+func activateAndSetupConda(projName string) error {
+	err := lib.CreateFile("tmp.sh")
+	defer os.Remove("tmp.sh")
+	if err != nil {
+		return err
+	}
+	tmp, err := os.OpenFile("tmp.sh", os.O_APPEND|os.O_WRONLY, 0644)
+	defer tmp.Close()
+	_, err = tmp.WriteString("eval \"$(conda shell.bash hook)\"\nconda activate " + projName + "\nconda install python\nconda install pip")
+	activate := exec.Command("sh", "tmp.sh")
+	activate.Stderr = os.Stderr
+	activate.Stdout = os.Stdout
+	err = activate.Run()
+	return err
+}
+
+// createInstallSh creates the install_packages.sh file with the shebang and
+// example comment showing how to use it to install packages with pip
+// * tested
+func createInstallSh() error {
+	err := lib.CreateFile("install_packages.sh")
+	if err != nil {
+		return err
+	}
+	install, err := os.OpenFile("install_packages.sh", os.O_APPEND|os.O_WRONLY, 0644)
+	defer func() {
+		err := install.Close()
+		lib.Check(err)
+	}()
+	_, err = install.WriteString("#!/bin/sh\n\n# Ex: pip install [package]==[version]\n")
+	return err
+}
+
+// createREADME creates a README.md file, with optional conda setup information
+// if conda is true
+// * tested
+func createREADME(projName string, conda bool) error {
+	err := lib.CreateFile("README.md")
+	if err != nil {
+		return err
+	}
+	readme, err := os.OpenFile("README.md", os.O_APPEND|os.O_WRONLY, 0644)
+	defer func() {
+		err := readme.Close()
+		lib.Check(err)
+	}()
+	header := "# " + projName
+	setup := "## Setup and Installation\n"
+	if conda {
+		setup += "Create a conda environment\n`conda create " + projName + "`\nActivate conda environment\n`conda activate " + projName + "`\nInstall python\n`conda install python`\nInstall pip\n`conda install pip`\n"
+	}
+	install := "Install Packages\n`sh install_packages.sh`\n\n"
+	canaveral := "###### Created with [Canaveral](https://github.com/jchengjr77/canaveral)"
+	_, err = readme.WriteString(header + "\n\n" + setup + install + canaveral)
+	return err
+}
+
+// AddPythonProj launches a new python project.
+// It will create a conda environment if conda is installed,
```

Comment:
What's the reason for choosing conda? I'm not a python expert, but I thought conda has a ton of problems when installed with node on the same computer?

Reply:
So, I've never actually had a problem with node and conda together, but regardless I think these problems mostly arise when you improperly install them together, not so much when you just have them together. If problems crop up down the road we can refactor.

But the reason I setup python projects in conda environments is because different projects can have different dependencies, and so when you install and delete all of your python dependencies at the global level, it can get very messy and things tend to mess with eachother. Conda environments handle this by basically installing them in different environments, so dependencies don't interact with eachother. This makes it really convenient to recreate someone's python environment, similarly to how docker works.

All of that said, I think it's kinda beside the point because conda is totally optional, all of the functionality will still work without it (although canaveral doesn't install python and pip for you if you don't have it. This is only done in the conda environment because in a new conda env they won't be there and they are necessary). So someone who uses canaveral and never does node development can happily make their conda envs, and someone who does node development but never uses python can have their npm. Then someone who uses both can deal with potential compatibility issues however they would if they weren't using canaveral, whether that's not using conda, not using npm, or something else.

Hope that clarifies it.

Reply:
Okay, so if I understand correctly, conda is essentially a partitioning tool? Kind of like docker, you said?

That makes sense to me. I do remember python environments being quite important.